File: app/src/main/java/in/sudopk/vaishnavacalendar/LocationFragment.java
package in.sudopk.vaishnavacalendar;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AppCompatDialogFragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import java.util.List;

import in.sudopk.coreandroid.Fm;
import in.sudopk.coreandroid.Layout;
import in.sudopk.vaishnavacalendar.retrofit.VcService;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LocationFragment extends AppCompatDialogFragment implements LocationContainer {
    public static final String TAG = LocationFragment.class.getName();
    private LocationAdapter mAdapter;
    private ProgressBar mProgressBar;
    private VcService mVcService;
    private LocationStore mLocationStore;
    private CalendarStore mCalendarStore;
    private RecyclerView mRecyclerView;

    public static DialogFragment newInstance() {
        return new LocationFragment();
    }

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setHasOptionsMenu(true);

        mVcService = ((VcApp) getActivity().getApplication()).getVcService();

        mLocationStore = ((VcApp) getActivity().getApplication()).getLocationStore();
        mCalendarStore = ((VcApp) getActivity().getApplication()).getCalendarStore();

    }

    @NonNull
    @Override
    public Dialog onCreateDialog(final Bundle savedInstanceState) {
        final Dialog dialog = super.onCreateDialog(savedInstanceState);
        dialog.setTitle(R.string.location_title);
        return dialog;
    }

    @Nullable
    @Override
    public View onCreateView(final LayoutInflater inflater,
                             @Nullable final ViewGroup container,
                             @Nullable final Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.location, container, false);
        mProgressBar = Layout.findViewById(view, R.id.progressBar);
        mRecyclerView = Layout.findViewById(view, R.id.recyclerView);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mAdapter = new LocationAdapter(this);
        mRecyclerView.setAdapter(mAdapter);
        return view;
    }

    @Override
    public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
        inflater.inflate(R.menu.calendar, menu);
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        if (item.getItemId() == R.id.refresh) {
            refresh();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        refresh();
    }

    @Override
    public void onCancel(final DialogInterface dialog) {
        super.onCancel(dialog);
        getContainer().onLocationSelectCanceled();
    }

    @Override
    public void onLocationSelected(final Location location) {
        dismiss();
        getContainer().onLocationSelected(location);
    }

    @Override
    public void onLocationSelectCanceled() {
        dismiss();
        getContainer().onLocationSelectCanceled();
    }

    private LocationContainer getContainer() {
        return Fm.container(this);
    }

    private void refresh() {
        if (isResumed()) {
            mProgressBar.setVisibility(View.VISIBLE);

            if (mLocationStore.hasLocations()) {
                onLocationResponse(mLocationStore.getLocations());
            } else {
                Call<List<Country>> locationsCall = mVcService.locations();
                locationsCall.enqueue(new Callback<List<Country>>() {
                    @Override
                    public void onResponse(final Call<List<Country>> call, final Response<List<Country>> response) {
                        if (isResumed()) {
                            if (response.body() != null) {
                                mLocationStore.saveLocations(response.body());
                                onLocationResponse(mLocationStore.getLocations());
                            } else {
                                failed();
                            }
                        }
                    }

                    @Override
                    public void onFailure(final Call<List<Country>> call, final Throwable t) {
                        t.printStackTrace();
                        failed();
                    }
                });
            }
        }
    }

    private void failed() {
        if (isResumed()) {
            mProgressBar.setVisibility(View.GONE);
            if (getView() != null) {
                Snackbar.make(getView(), R.string.locations_failure, Snackbar.LENGTH_SHORT).show();
            }
        }
    }

    private void onLocationResponse(final List<Country> response) {
        if (isResumed()) {
            mProgressBar.setVisibility(View.GONE);
            mAdapter.setData(response);

            int position = mAdapter.getPosition(mCalendarStore.getLocation());
            if (position > 0) {
                // lets scroll to location one before the actual one, looks little better
                position--;
            }
            mRecyclerView.getLayoutManager()
                    .scrollToPosition(position);
        }
    }

}

File: app/src/main/java/in/sudopk/vaishnavacalendar/location/LocationFragment.kt
package `in`.sudopk.vaishnavacalendar.location

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.DialogFragment
import android.support.v7.app.AppCompatDialogFragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.ProgressBar

import `in`.sudopk.coreandroid.Fm
import `in`.sudopk.coreandroid.Layout
import `in`.sudopk.vaishnavacalendar.R
import `in`.sudopk.vaishnavacalendar.VcApp
import `in`.sudopk.vaishnavacalendar.calendar.CalendarStore
import `in`.sudopk.vaishnavacalendar.calendar.Country
import `in`.sudopk.vaishnavacalendar.retrofit.VcService
import `in`.sudopk.vaishnavacalendar.vcApp
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class LocationFragment : AppCompatDialogFragment(), LocationContainer {
    private lateinit var mAdapter: LocationAdapter
    private lateinit var mProgressBar: ProgressBar
    private lateinit var mVcService: VcService
    private lateinit var mLocationStore: LocationStore
    private lateinit var mCalendarStore: CalendarStore
    private lateinit var mRecyclerView: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //setHasOptionsMenu(true);

        mVcService = vcApp.vcService

        mLocationStore = vcApp.locationStore
        mCalendarStore = vcApp.calendarStore

        setStyle(DialogFragment.STYLE_NORMAL, R.style.Theme_AppCompat_Light_Dialog)

    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        dialog.setTitle(R.string.location_title)
        return dialog
    }

    override fun onCreateView(inflater: LayoutInflater?,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.location, container, false)
        mProgressBar = Layout.findViewById<ProgressBar>(view, R.id.progressBar)
        mRecyclerView = Layout.findViewById<RecyclerView>(view, R.id.recyclerView)
        mRecyclerView.layoutManager = LinearLayoutManager(context)
        mAdapter = LocationAdapter(this, mCalendarStore.location)
        mRecyclerView.adapter = mAdapter
        return view
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.calendar, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item!!.itemId == R.id.refresh) {
            refresh()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        refresh()
    }

    override fun onCancel(dialog: DialogInterface?) {
        super.onCancel(dialog)
        container.onLocationSelectCanceled()
    }

    override fun onLocationSelected(location: Location) {
        dismiss()
        container.onLocationSelected(location)
    }

    override fun onLocationSelectCanceled() {
        dismiss()
        container.onLocationSelectCanceled()
    }

    private val container: LocationContainer
        get() = Fm.container<LocationContainer>(this)

    private fun refresh() {
        if (isResumed) {
            mProgressBar.visibility = View.VISIBLE

            if (mLocationStore.hasLocations()) {
                onLocationResponse(mLocationStore.locations)
            } else {
                val locationsCall = mVcService.locations()
                locationsCall.enqueue(object : Callback<List<Country>> {
                    override fun onResponse(call: Call<List<Country>>, response: Response<List<Country>>) {
                        if (isResumed) {
                            if (response.body() != null) {
                                mLocationStore.saveLocations(response.body())
                                onLocationResponse(mLocationStore.locations)
                            } else {
                                failed()
                            }
                        }
                    }

                    override fun onFailure(call: Call<List<Country>>, t: Throwable) {
                        t.printStackTrace()
                        failed()
                    }
                })
            }
        }
    }

    private fun failed() {
        if (isResumed) {
            mProgressBar.visibility = View.GONE
            val v = view
            if (v != null) {
                Snackbar.make(v, R.string.locations_failure, Snackbar.LENGTH_SHORT).show()
            }
        }
    }

    private fun onLocationResponse(response: List<Country>) {
        if (isResumed) {
            mProgressBar.visibility = View.GONE
            mAdapter.setData(response)

            var position = mAdapter.getPosition(mCalendarStore.location)
            if (position > 0) {
                // lets scroll to location one before the actual one, looks little better
                position--
            }
            mRecyclerView.layoutManager
                    .scrollToPosition(position)
        }
    }

    companion object {
        val TAG = LocationFragment::class.java.name

        fun newInstance(): DialogFragment {
            return LocationFragment()
        }
    }

}
